hq/app/services/MetricService.scala

```diff
@@ -35,10 +35,15 @@ class MetricService(
       return
     }
 
-    Cloudwatch.logAsMetric(allSgs, Cloudwatch.DataType.sgTotal)
-    Cloudwatch.logAsMetric(allExposedKeys, Cloudwatch.DataType.iamKeysTotal)
-    Cloudwatch.logAsMetric(allPublicBuckets, Cloudwatch.DataType.s3Total)
-    Cloudwatch.logMetricsForCredentialsReport(allCredentials)
+    for {
+      allGcpFindings <- cacheService.getGcpFindings
+    } yield {
+      Cloudwatch.logAsMetric(allSgs, Cloudwatch.DataType.sgTotal)
+      Cloudwatch.logAsMetric(allExposedKeys, Cloudwatch.DataType.iamKeysTotal)
+      Cloudwatch.logAsMetric(allPublicBuckets, Cloudwatch.DataType.s3Total)
+      Cloudwatch.logMetricsForCredentialsReport(allCredentials)
```

Comment:
I think by moving these calls inside the for/yield, it now means that if `getGcpFindings` fails then nothing in the yield will run. It might be best to keep them outside (assuming I understand this correctly!)

Reply:
So that's actually intended behaviour. By only sending complete datasets, it makes it possible to aggregate them over time in a meaningful way.

I'm not sure what happens if the GCP call fails. Will it throw an exception? Does this need a try block?

Reply:
Is that still desired behaviour even for GCP where we aren't going to be able to correlate the accounts against one another?

The GCP call returns an either so it won't fail - your for comprehension just ignores the failure case.

Reply:
`> The GCP call returns an either so it won't fail - your for comprehension just ignores the failure case.`
Ahhh, got it!

`> Is that still desired behaviour even for GCP where we aren't going to be able to correlate the accounts against one another?`

Yeah, you're right, I'm pushing for something here that is quite counter intuitive and maybe I need to abandon it. Here's my thinking on this, you tell me if I'm overcomplicating it

So, to answer your question yes - this makes it easy to calculate a total number of vulnerabilities. Say for example we do a query like `Vulnerabilities where DataType contains '/critical'` and we get

| Time | DataType  | Vulnerabilities |
| -------------| ------------- | ------------- |
| 1pm | iam/credentials/critical  | 5  |
| 1pm | gcp/critical  | 3 |
| 2pm | iam/credentials/critical  | 5  |
| 3pm | iam/credentials/critical  | 5  |
| 3pm | gcp/critical  | 2 |
| 4pm | iam/credentials/critical  | 5  |
| 4pm | gcp/critical  | 3 |

If we do a sum aggregated per time we get

| Time | CriticalVulnerabilities |
| -------------| ------------- |
| 1pm | 8  |
| 2pm | 5  |
| 3pm | 7 |
| 4pm | 8 |

The 2pm value is not meaningful.

There is a workaround for this, which is to aggregate on a longer period. This only works if our data _has_ a period that guarantees we'll find a value for every DataType.

In the example it's pretty easy. We could aggregate on day instead.

| Time | CriticalVulnerabilities |
| -------------| ------------- |
| Today | 8... or maybe 7? |

When doing a roll up like this, we can pick one of 6 statistical operations. We can pick for example the minimum (7), maximum (8), average (~7.5), but [there's no function to pick the most recent value](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_concepts.html#Statistic), which is annoying. So you can't aggregate this in a way where cloudwatch will add the 4pm gcp/critical with the 4pm /iam/credentials/critical values.

So I guess my pitch is that by sacrificing resolution we get a dataset that's easier to work with. So far this hasn't been problematic, ever since we fixed the rate limiting issue the data is consistent. Do you anticipate GCP will be flaky?

Reply:
You make an excellent case. I think you're right and we should do all or nothing - maybe in the future we might alert off 'critical vulnerabilities >0' and it's important that is based off complete data. Maybe just add a comment here to explain why the other metrics are dependent on GCP - could even link directly to this discussion